r/R/dplyr-functions.R

```diff
@@ -215,6 +215,52 @@ nse_funcs$nchar <- function(x, type = "chars", allowNA = FALSE, keepNA = NA) {
   }
 }
 
+nse_funcs$paste <- function(..., sep = " ", collapse = NULL, recycle0 = FALSE) {
+  assert_that(is.null(collapse))
+  arrow_string_join_function(NullHandlingBehavior$REPLACE, "NA")(..., sep)
+}
+
+nse_funcs$paste0 <- function(..., collapse = NULL, recycle0 = FALSE) {
+  assert_that(is.null(collapse))
+  arrow_string_join_function(NullHandlingBehavior$REPLACE, "NA")(..., "")
+}
+
+nse_funcs$str_c <- function(..., sep = "", collapse = NULL) {
+  assert_that(is.null(collapse))
+  arrow_string_join_function(NullHandlingBehavior$EMIT_NULL)(..., sep)
+}
+
+arrow_string_join_function <- function(null_handling, null_replacement = NULL) {
+  # the binary_join_element_wise kernel takes the separator as the last argument
+  function(...) {
+    dots <- list(...) # sep is the last value in dots
+    for (i in seq_along(dots)) {
+      # handle scalar literal args, and cast all args to string for
+      # consistency with base::paste(), base::paste0(), and stringr::str_c()
+      if (!inherits(dots[[i]], "Expression")) {
+        assert_that(
+          length(dots[[i]]) == 1,
+          msg = "Literal vectors of length != 1 not supported in string concatenation"
+        )
+        # handle scalar literal NA consistent with the binary_join_element_wise
+        # kernel's handling of nulls in the data
+        if (null_handling == NullHandlingBehavior$REPLACE && is.na(dots[[i]])) {
+          dots[[i]] <- null_replacement
+        }
```

Comment:
This is necessary (for reasons I don't fully understand) to enable use of `NA` literal scalars in the `paste()`, `paste0()`, and `str_c()` expressions. See the discussion about this at https://github.com/apache/arrow/pull/10520#issuecomment-862962267